dace/transformation/interstate/block_fusion.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2019-2024 ETH Zurich and the DaCe authors. All rights reserved.
+
+from dace.sdfg import utils as sdutil
+from dace.sdfg.state import AbstractControlFlowRegion, ControlFlowBlock, ControlFlowRegion, SDFGState
+from dace.transformation import transformation
+
+
+@transformation.experimental_cfg_block_compatible
+class BlockFusion(transformation.MultiStateTransformation):
+    """ Implements the state-fusion transformation.
+
+        State-fusion takes two states that are connected through a single edge,
+        and fuses them into one state. If permissive, also applies if potential memory
+        access hazards are created.
+    """
+
+    first_block = transformation.PatternNode(ControlFlowBlock)
+    second_block = transformation.PatternNode(ControlFlowBlock)
+
+    @staticmethod
+    def annotates_memlets():
+        return False
+
+    @classmethod
+    def expressions(cls):
+        return [sdutil.node_path_graph(cls.first_block, cls.second_block)]
+
+    def _is_noop(self, block: ControlFlowBlock) -> bool:
+        if isinstance(block, SDFGState):
+            return block.is_empty()
+        elif type(block) == ControlFlowBlock:
+            return True
+        return False
+
+    def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
+        # First block must have only one unconditional output edge (with dst the second block).
+        out_edges = graph.out_edges(self.first_block)
+        if len(out_edges) != 1 or out_edges[0].dst is not self.second_block or not out_edges[0].data.is_unconditional():
+            return False
+        # Inversely, the second block may only have one input edge, with src being the first block.
+        in_edges_second = graph.in_edges(self.second_block)
+        if len(in_edges_second) != 1 or in_edges_second[0].src is not self.first_block:
+            return False
+
+        # Ensure that either that both blocks are fusable blocks, meaning that at least one of the two blocks must be
+        # a 'no-op' block. That can be an empty SDFGState or a general control flow block without further semantics
+        # (no loop, conditional, break, continue, control flow region, etc.).
+        if not self._is_noop(self.first_block) and not self._is_noop(self.second_block):
+            return False
+
+        # The interstate edge may have assignments if there are input edges to the first block that can absorb them.
+        in_edges = graph.in_edges(self.first_block)
+        if out_edges[0].data.assignments:
+            if not in_edges:
+                return False
+            # If the first block is a control flow region, no absorbtion is possible.
+            if isinstance(self.first_block, AbstractControlFlowRegion):
```

Comment:
Not quite, `SDFGState`s are not instances of control flow regions and would not match that anyway. But both all types of control flow regions (collectively subclasses of `AbstractControlFlowRegion`) and SDFGStates, as well as any other control flow blocks, such as `BreakBlock`, `ContinueBlock`, or `ReturnBlock` are all subclasses of `ControlFlowBlocks`. So what this check is testing, is if the block is a basic block (states, continue, return, etc.), or any kind of control flow region (anything under `AbstractControlFlowRegion`). Testing just for `ControlFlowRegion` would exclude `ConditionalBlocks`.

Reply:
We need a readthedocs page on the class structure of SDFG contents. This can be a bit confusing, especially with conditional blocks. This can be a separate PR

Reply:
Tracking in https://github.com/orgs/spcl/projects/16